RFC: Raising zero to the power of negative number

jorgsowa opened the pull request:
Fixes: https://github.com/php/php-src/issues/8015

RFC: https://wiki.php.net/rfc/raising_zero_to_power_of_negative_number

Based on the code: https://github.com/php/php-src/compare/master...iluuu1994:php-src:throw-division-by-zero-on-pow-of-base-0-exponent-lt-eq-0

iluuu1994 commented on Zend/zend_operators.c:
```diff
@@ -1287,6 +1287,21 @@ ZEND_API zend_result ZEND_FASTCALL mul_function(zval *result, zval *op1, zval *o
+static zend_result safe_pow(double *result, double base, double exponent)
+{
+	if (base == 0.0 && exponent < 0.0) {
+		zend_power_base_0_exponent_lt_eq_0_error();
```
This should likely check for `EG(exception)` and return `FAILURE` in that case (for throwing error handlers). Usages of `safe_pow` should propagate this return value.

jorgsowa replied:
I'm not sure what I should change here. The changes should not throw an exception but the deprecation note. In PHP 9.0 it should throw an exception, but this PR is aimed at 8.4. The same note applies to your second comment.

iluuu1994 replied:
Warnings can be promoted from error handlers to exceptions. In these cases, often, the program flow must be adjusted in some way. Anyway, I looked at the way `pow_function_base()`s return value is used, and apparently it indicates not whether the operation was successful, but whether the two operand types support the `pow` operation. So a change is probably not necessary here.

iluuu1994 commented on Zend/zend_operators.c:
```diff
@@ -1372,7 +1399,7 @@ ZEND_API zend_result ZEND_FASTCALL pow_function(zval *result, zval *op1, zval *o
 		return SUCCESS;
 	}
 
-	ZEND_ASSERT(0 && "Operation must succeed");
```
I don't think this needs to change?

jorgsowa replied:
Thx. Removed

iluuu1994 commented on Zend/zend_compile.c:
```diff
@@ -8902,6 +8902,10 @@ ZEND_API bool zend_binary_op_produces_error(uint32_t opcode, const zval *op1, co
 	}
+	if ((opcode == ZEND_POW) && zval_get_long(op1) == 0 && zval_get_long(op2) < 0) {
+		/* 0 ** (<=0) throws a division by zero error. */
+		zend_error(E_DEPRECATED, "Power of base 0 and negative exponent is deprecated");
```
This function shouldn't error itself. It should indicate whether the operation would error at runtime. In that case, it is not executed at compile time, so that the error may be caught at runtime. Hence, you can just return `true` here.

iluuu1994 commented:
@jorgsowa I'm assuming this is ready to have another look?

iluuu1994 commented:
Seems like the comments are still unaddressed. Please ping me once you've made the relevant changes, or if you need help with them.

jorgsowa commented:
Thanks for review @iluuu1994. I changed the status to Draft to not take the time of the reviewers when the PR is not finished. Now I have added the function `fpow` and addressed all comments. I will still wait for the tests to pass until I change it back to the version to review.

iluuu1994 commented on Zend/zend_operators.c:
```diff
@@ -1372,7 +1399,7 @@ ZEND_API zend_result ZEND_FASTCALL pow_function(zval *result, zval *op1, zval *o
 	}
 
-	ZEND_ASSERT(0 && "Operation must succeed");
+    ZEND_ASSERT(0 && "Operation must succeed");
```
Nit: Whitespace was changed.

iluuu1994 commented on Zend/zend_compile.c:
```diff
@@ -8978,6 +8978,10 @@ ZEND_API bool zend_binary_op_produces_error(uint32_t opcode, const zval *op1, co
 		/* Division by zero throws an error. */
 		return 1;
 	}
+	if ((opcode == ZEND_POW) && zval_get_long(op1) == 0 && zval_get_long(op2) < 0) {
```
`<` should be `<=`

jorgsowa replied:
I fixed the comment below. It applies to negative exponents, not 0.

TimWolla replied:
FWIW: It was not part of the RFC, but `0` raised to `0` would also be undefined.

iluuu1994 replied:
> FWIW: It was not part of the RFC, but 0 raised to 0 would also be undefined.

I also wondered about this. But apparently it is not universally considered undefined. See https://en.wikipedia.org/wiki/Zero_to_the_power_of_zero.

jorgsowa replied:
According to [IEEE 754-2008](https://en.wikipedia.org/wiki/Zero_to_the_power_of_zero#IEEE_floating-point_standard)

> pow (x, ±0) is 1 for any x (even a zero, quiet NaN, or infinity)

There is no specialized `powr()` function in PHP which treats 0**0 as NaN.

TimWolla replied:
> According to IEEE 754-2008

For `fpow()` that absolutely makes sense. However for `pow()` and `**`, PHP could opt for a different behavior that would throw on the undefined operation. Like division by zero is also well-defined for `fdiv()`, but throws with `/`.

jorgsowa replied:
That's true. Although it's not so unambiguous. We would need to create another RFC for this.

TimWolla replied:
Yes, that would not be covered this RFC. I primarily wanted to note it for posterity, because it is related.

iluuu1994 commented on Zend/zend_operators.c:
```diff
@@ -1287,6 +1287,21 @@ ZEND_API zend_result ZEND_FASTCALL mul_function(zval *result, zval *op1, zval *o
+
+static zend_result safe_pow(double *result, double base, double exponent)
+{
+	if (base == 0.0 && exponent < 0.0) {
```
You could wrap the condition in `UNEXPECTED`.

TimWolla commented on Zend/zend_operators.c:
```diff
@@ -1287,6 +1287,21 @@ ZEND_API zend_result ZEND_FASTCALL mul_function(zval *result, zval *op1, zval *o
+
+static zend_result safe_pow(double *result, double base, double exponent)
+{
+	if (UNEXPECTED(base == 0.0 && exponent < 0.0)) {
+		zend_power_base_0_exponent_lt_eq_0_error();
```
Function name is inconsistent with the condition (< vs lt_eq)

jorgsowa replied:
Thanks. Renamed function.

jorgsowa commented on Zend/zend_operators.c:
```diff
@@ -1287,6 +1287,21 @@ ZEND_API zend_result ZEND_FASTCALL mul_function(zval *result, zval *op1, zval *o
+	zend_error(E_DEPRECATED, "Power of base 0 and negative exponent is deprecated");
+}
+
+static zend_result safe_pow(double *result, double base, double exponent)
```
Sure. I changed the return type of this function.

iluuu1994 commented on ext/standard/tests/math/is_finite_basic.phpt:
```diff
@@ -2,26 +2,29 @@
 ?>
---EXPECT--
+--EXPECTF--
+
```
Nit: Whitespace can be removed. I believe the output is trimmed anyway.

jorgsowa replied:
Removed

iluuu1994 reviewed (approved):
This looks good to me. Thanks for the changes! You can ignore the nit, I'll adjust it before merging.

jorgsowa commented on ext/standard/tests/math/is_finite_basic.phpt:
```diff
@@ -2,26 +2,29 @@
-                pow(0, -2),
-                acos(1.01));
-;
+$values = [
+    234,
+    -234,
+    23.45e1,
+    -23.45e1,
+    0xea,
+    0352,
+    "234",
+    "234.5",
+    "23.45e1",
+    true,
+    false,
+    pow(0, -2),
+    acos(1.01),
+];
```
That's true. I reformatted the code as I have already touched the file within PR.

TimWolla replied:
Unrelated changes, such as this reformatting, make it harder to see what changed in the test over time, e.g. within the `git blame` output: As a reader I would need to think whether the changes are significant or not. As an example, this change makes it harder to determine if this RFC added a new test case to properly cover the changed code.

jorgsowa replied:
Do you think it would be good to reformat all such tests in a separate PR? I also had some cases in the past when I was modifying existing tests and the code was not properly formatted so I just included this in PR. If we reformat everything at once then we don't have such discussions again.

iluuu1994 replied:
I'm generally against reformatting, but this formatting is just objectively bad. So, doing it in a separate PR makes sense to me.

jorgsowa commented:
I added an entry about deprecation to `NEWS`.

TimWolla commented:
> I added an entry about deprecation to `NEWS`.

Either both NEWS and UPGRADING should be adjusted, or neither of them. Within NEWS, the [RFC] tag is not used historically and looks odd.

jorgsowa commented:
@TimWolla thank your comments. I adjusted the entry in NEWS. I was following the existing tag in the file but changed to the more common convention. Also, I added an entry to the UPGRADING file.

iluuu1994 commented:
Thanks @jorgsowa! Merged without the formatting changes. I also fixed a small bug in `zend_binary_op_produces_error()`. Feel free to mark the RFC as implemented. Please also adjust your editor settings to use tabs instead of spaces for C files. These are easy to miss in reviews (I don't use the Refined GitHub extension because it lags).

jorgsowa commented:
Thank you for your help with this PR. I will prepare PR with the reformatting of those test snippets and will change my code style configuration.
